crates/hyperdrive-math/src/lp.rs

```diff
@@ -600,6 +621,567 @@ impl State {
+        active_lp_total_supply: FixedPoint,
+        withdrawal_shares_total_supply: FixedPoint, // withdraw shares - ready to withdraw
+        max_iterations: u64,
+    ) -> Result<(FixedPoint, FixedPoint)> {
+        // Steps 1 and 2: Calculate the maximum amount the share reserves can be
+        // debited. If the effective share reserves or the maximum share
+        // reserves delta can't be calculated or if the maximum share reserves
+        // delta is zero, idle can't be distributed.
+        let max_share_reserves_delta =
+            self.calculate_max_share_reserves_delta(current_block_timestamp)?;
+
+        if max_share_reserves_delta == fixed!(0) {
+            return Err(eyre!("Max share reserves delta is zero."));
+        }
+
+        // Step 3: Calculate the amount of withdrawal shares that can be
+        // redeemed given the maximum share reserves delta.  Otherwise, we
+        // proceed to calculating the amount of shares that should be paid out
+        // to redeem all of the withdrawal shares.
+        let withdrawal_shares_redeemed = {
+            let withdrawal_shares_redeemed = self
+                .calculate_distribute_excess_idle_withdrawal_shares_redeemed(
+                    current_block_timestamp,
+                    max_share_reserves_delta,
+                    active_lp_total_supply,
+                    withdrawal_shares_total_supply,
+                )?;
+
+            // Step 3: If none of the withdrawal shares could be redeemed, then
+            // we're done and we pay out nothing.
+            if withdrawal_shares_redeemed == fixed!(0) {
+                return Err(eyre!(
+                    "Withdrawal shares redeemed from excess idle is zero."
+                ));
+            }
+            // Step 3: Otherwise if this amount is less than or equal to the amount
+            // of withdrawal shares outstanding, then we're done and we pay out the
+            // full maximum share reserves delta.
+            else if withdrawal_shares_redeemed <= withdrawal_shares_total_supply {
+                return Ok((withdrawal_shares_redeemed, max_share_reserves_delta));
+            }
+            // Step 3: Otherwise, all of the withdrawal shares are redeemed, and we
+            // need to calculate the amount of shares the withdrawal pool should
+            // receive.
+            else {
+                withdrawal_shares_total_supply
+            }
+        };
+
+        // Step 4: Solve for the share proceeds that hold the LP share price
+        // invariant after all of the withdrawal shares are redeemed. If the
+        // calculation returns a share proceeds of zero, we can't pay out
+        // anything.
+        let share_proceeds = self.calculate_distribute_excess_idle_share_proceeds(
+            current_block_timestamp,
+            active_lp_total_supply,
+            withdrawal_shares_total_supply,
+            max_share_reserves_delta,
+            max_iterations,
+        )?;
```

Comment:
Double-checking this, this `?` looks correct since `calculate_distribute_excess_idle_share_proceeds` isn't "safe" in Solidity: https://github.com/delvtech/hyperdrive/blob/main/contracts/src/libraries/LPMath.sol#L624. With this in mind, we'll need to be more careful about the values that we return inside of this function call. Most of the time, it returns zero when an underlying "safe" function fails, but there are a few places where returning an `Err` is appropriate (like with `I256::try_from`).